awx/sso/social_pipeline.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) 2015 Ansible, Inc.
+# All Rights Reserved.
+
+# Python
+import re
+import logging
+
+from awx.sso.common import get_or_create_with_default_galaxy_cred
+
+logger = logging.getLogger('awx.sso.social_pipeline')
+
+
+def _update_m2m_from_expression(user, related, expr, remove=True):
+    """
+    Helper function to update m2m relationship based on user matching one or
+    more expressions.
+    """
+    should_add = False
+    if expr is None:
+        return
+    elif not expr:
+        pass
+    elif expr is True:
+        should_add = True
+    else:
+        if isinstance(expr, (str, type(re.compile('')))):
+            expr = [expr]
+        for ex in expr:
+            if isinstance(ex, str):
+                if user.username == ex or user.email == ex:
+                    should_add = True
+            elif isinstance(ex, type(re.compile(''))):
+                if ex.match(user.username) or ex.match(user.email):
+                    should_add = True
+    if should_add:
+        related.add(user)
+    elif remove:
+        related.remove(user)
+
+
+def update_user_orgs(backend, details, user=None, *args, **kwargs):
+    """
+    Update organization memberships for the given user based on mapping rules
+    defined in settings.
+    """
+    if not user:
+        return
+
+    org_map = backend.setting('ORGANIZATION_MAP') or {}
+    for org_name, org_opts in org_map.items():
+        organization_alias = org_opts.get('organization_alias')
+        if organization_alias:
+            organization_name = organization_alias
+        else:
+            organization_name = org_name
+        org = get_or_create_with_default_galaxy_cred(name=organization_name)
```

Comment:
huh, I'm surprised that we don't use `defaults` already, like, set the description of the org to "imported via SAML"

Reply:
Its just a straight building of the org. This will be a slight change to how the LDAP adapter works because it didn't used to include adding the default galaxy credential at all.